### ISSUE TITLE
visit unregistered websocket handler will crash civetweb

### ISSUE THREAD

**Issue by quhezheng** [civetweb/civetweb#901]
version 1.11  

Register WS handler by CivetServer::AddWebsocketHandler() is OK, visit the WB handler is OK. 
But if visit the WRONG WS address, civetweb will crash, call stack:

handle_websocket_request.isra
handle_request
process_new_connection
worker_thread
start_thread
clone

**Comment by bel2125** [comment]
> But if visit the WRONG WS address, civetweb will crash

What do you mean by "visit the wrong websocket address"? You try to open a websocket connection to an ws://host/mywebsock URL, where there is no handler for "/mywebsock" registered? There never was an handler for this URL *or* there was a handler but you unregistered it?

**Comment by quhezheng** [comment]
> > But if visit the WRONG WS address, civetweb will crash
> 
> What do you mean by "visit the wrong websocket address"? You try to open a websocket connection to an ws://host/mywebsock URL, where there is no handler for "/mywebsock" registered? There never was an handler for this URL _or_ there was a handler but you unregistered it?


Tried to open a websocket connection to an ws://host/mywebsock URL, but there is no handler for "/mywebsock" registered.

**Comment by bel2125** [comment]
Just checked with the current HEAD version as well as with V1.11, but did not see any crash. 
I used the same websocket.xhtml test file as in the test/ directory, but changed the websocket URL to something non-existing.
Could you provide some steps to reproduce this issue?

**Comment by quhezheng** [comment]
I played the trick during the compiling process to add 'MG_LEGACY_INTERFACE'
I editded include/civetweb.h by adding following to line 25

#ifndef MG_LEGACY_INTERFACE
#define MG_LEGACY_INTERFACE
#endif

make lib WITH_CPP=1 WITH_WEBSOCKET=1 WITH_ZLIB=1

**Comment by bel2125** [comment]
I added the MG_LEGACY_INTERFACE, fixed some compile issues (old LEGAGY_INTERFACEs are untested), and successfully tested it with embedded_cpp:  
```
make lib WITH_CPP=1 WITH_WEBSOCKET=1 WITH_ZLIB=1
cd examples/embedded_cpp
make WITH_CPP=1 WITH_WEBSOCKET=1 WITH_ZLIB=1
cp ../../test/websocket_wrong_url.xhtml .
./embedded_cpp
--> open browser: http://localhost:8081/websocket_wrong_url.xhtml
```
No crash.
I cannot help as long as I cannot reproduce the problem.

Probably you could turn off optimization, add debug information (use a debug build) to get line numbers.
The "*.isra" extension in your call stack tells my, you are using at least -O2 (see [here](https://stackoverflow.com/questions/18907580/what-is-isra-in-the-kernel-thread-dump)). If you replace -O2 (-Os, -O3) by "-Og -g -ggdb", you should get line numbers in the calls stack, and if you get a core file even values of variables at the crash.

**Comment by quhezheng** [comment]
> I added the MG_LEGACY_INTERFACE, fixed some compile issues (old LEGAGY_INTERFACEs are untested), and successfully tested it with embedded_cpp:
> 
> ```
> make lib WITH_CPP=1 WITH_WEBSOCKET=1 WITH_ZLIB=1
> cd examples/embedded_cpp
> make WITH_CPP=1 WITH_WEBSOCKET=1 WITH_ZLIB=1
> cp ../../test/websocket_wrong_url.xhtml .
> ./embedded_cpp
> --> open browser: http://localhost:8081/websocket_wrong_url.xhtml
> ```
> 
> No crash.
> I cannot help as long as I cannot reproduce the problem.
> 
> Probably you could turn off optimization, add debug information (use a debug build) to get line numbers.
> The "*.isra" extension in your call stack tells my, you are using at least -O2 (see [here](https://stackoverflow.com/questions/18907580/what-is-isra-in-the-kernel-thread-dump)). If you replace -O2 (-Os, -O3) by "-Og -g -ggdb", you should get line numbers in the calls stack, and if you get a core file even values of variables at the crash.

I walked around the crash by bypassing the block

> https://github.com/civetweb/civetweb/blob/ce8f6d38a60eb16c996afee1e5340f76ef4d0923/src/civetweb.c#L13965

changed to an undefined macro as below :
`#if defined(__MG_LEGACY_INTERFACE__)`

The crash is no longer anymore

**Comment by bel2125** [comment]
Interesting you ran into this `if` branch, while I did not in my test, that should do the same thing.

Anyway, this part of the code is declared "legacy" since August 19, 2015 - so this code should have been removed already, since the tests for this have already been removed. I will remove some of the code that has been declared legacy for 3+ years - this corresponds to the solution you have already found.
